src/code/breakpointManager.js

```diff
@@ -17,37 +18,88 @@ class breakpointManager {
     this.event = new EventManager()
     this.debugger = _debugger
     this.breakpoints = {}
-    this.isPlaying = false
     this.locationToRowConverter = _locationToRowConverter
-    this.currentLine
+    this.previousLine
   }
 
   /**
     * start looking for the next breakpoint
+    * @param {Bool} defaultToLimit - if true jump to the enf of the trace if no more breakpoint found
     *
     */
-  async jumpNextBreakpoint (defaultToEnd) {
-    this.isPlaying = true
+  async jumpNextBreakpoint (defaultToLimit) {
+    this.jump(1, defaultToLimit)
+  }
+
+  /**
+    * start looking for the previous breakpoint
+    * @param {Bool} defaultToLimit - if true jump to the enf of the trace if no more breakpoint found
+    *
+    */
+  async jumpPreviousBreakpoint (defaultToLimit) {
+    this.jump(-1, defaultToLimit)
+  }
+
+   /**
+    * start looking for the previous or next breakpoint
+    * @param {Int} direction - 1 or -1 direction of the search
+    * @param {Bool} defaultToLimit - if true jump to the enf of the trace if no more breakpoint found
+    *
+    */
+  async jump (direction, defaultToLimit) {
+    if (!this.locationToRowConverter) {
+      console.log('row converter not provided')
+      return
+    }
+
+    function hitLine (currentStep, sourceLocation, self) {
+      if (helper.isJumpDestInstruction(self.debugger.traceManager.trace[currentStep]) ||
```

Comment:
I think the condition should rather be:
If the previous step was a jump out (that is important, otherwise we might miss breakpoints set at the beginning of a loop) or the previous call stack depth was one higher (if you do not use call stack depth, you miss out of gas conditions, which could occur with any opcode).